### PR TITLE
Add subscription table to MaestroWeb

### PR DESCRIPTION
**Pull request by meganaquinn**
Add a Subscriptions table that shows the subscriptions for the selected repo, grouped by branch. Also displays if they're enabled or disabled.

ETA Current UI:
image

### CODE REVIEW

File: src/Maestro/maestro-angular/src/app/page/build/build.component.ts [hunk 1]
```diff
@@ -59,12 +60,15 @@ export class BuildComponent implements OnInit, OnChanges {
   public toastDate?: Date;
   public acceptToast?: () => void;
 
+  public includeSubToolsets: boolean = false;
```
Comment (adiaaida) [inline src/Maestro/maestro-angular/src/app/page/build/build.component.ts]: How is this different from the includeToolsets variable above? Does subscription need its own toolsets toggle?

File: src/Maestro/maestro-angular/src/app/page/build/build.component.ts [hunk 2]
```diff
@@ -210,24 +214,49 @@ export class BuildComponent implements OnInit, OnChanges {
...
+          if (!emittedLoading) {
             emittedLoading = true;
             return true;
           }
           return false;
         }),
       ),
     );
+
+    this.subscriptionsList$ = this.build$.pipe(
```
Comment (adiaaida) [inline src/Maestro/maestro-angular/src/app/page/build/build.component.ts]: We don't seem to use the includeSubToolsets variable when grabbing the list of subscriptions. Is that future work? Should we just include it in the original design? Does having a toolset toggle even make sense for subscriptions?

File: src/Maestro/maestro-angular/src/app/page/build/build.component.html [hunk 3]
```diff
@@ -121,9 +125,33 @@ <h5 class="card-title float-left">
         </ng-template>
       </div>
 
+      <div class="container-fluid">
+        <div class="card mt-3">
+          <div class="card-body">
+            <h5 class="card-title float-left">
+              Subscriptions
+            </h5>
+            <!-- <span class="form-check float-right">
```
Comment (adiaaida) [inline src/Maestro/maestro-angular/src/app/page/build/build.component.html]: Let's remove the commented out code if we aren't ready to enable this. I don't like the idea of checking in commented out code.

**Comment by adiaaida** [comment]
It looks like the issue with .Net Tools - Latest isn't you. It happens in prod too, for corefx, etc. When I experimented, I checked arcade in prod, which was fine, but not corefx or any of the other non-tool repos, which are not. I'll open a bug to track this.

File: src/Maestro/maestro-angular/src/app/page/subscriptions-table/subscriptions-table.component.html [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    </li>
+  </ng-container>
+</ul>
+<table class="table table-sm table-hover mb-0" style="white-space: nowrap;">
+  <thead>
+    <th>Source Repository</th>
+    <th>Enabled</th>
+  </thead>
+  <tbody>
+    <ng-container *ngFor="let item of subscriptionsList[currentTab]">
```
Comment (chcosta) [inline src/Maestro/maestro-angular/src/app/page/subscriptions-table/subscriptions-table.component.html]: This causes an error if `currentTab` is undefined

File: src/Maestro/maestro-angular/src/app/page/subscriptions-table/subscriptions-table.component.html [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+      <tr>
+        <td>
+          {{item.sourceRepository}}
```
Comment (chcosta) [inline src/Maestro/maestro-angular/src/app/page/subscriptions-table/subscriptions-table.component.html]: Rather than tabs for branches, it might be simpler if the "Source Repository" was merely "Source" and included something like "[repository]@[branch]".

**Comment by adiaaida** [comment]
So I've been thinking about this, and I want to make sure that we are listing the subscriptions in the right way. My understanding is that everything under a channel on the left hand drop down should be considered a source repo. Ie, BarViz tells us what builds for each repo are being published to a particular channel. We list the dependencies for each of these repositories already (ie the builds that came into that particular build of the current repository).

Should the subscription table tell us what subscriptions are *coming* into the repository from a particular channel or what repositories a subscription is *going* into for this particular channel? I ask this because I actually find the Subscription table a little confusing as it is because for, say, corefx in .Net Core 5 Dev, we see that we have a dependency on arcade, but it isn't listed in the subscriptions list because that dependency is coming from the .Net Tools - Latest channel. But because the list on the left hand side only works for repos that produce builds that publish to a channel, and corefx doesn't publish to .Net Tools - Latest, we can't see that there is a subscription between arcade and corefx.

So what I am proposing is that we flip how we are thinking about subscriptions, and that the subscription table should consider the current repo as the SourceRepository and list out the TargetRepositories (ie, where this product will flow to). Then, we would go to .Net Tools - Latest > arcade and see that arcade flows... well, everywhere. And find out if any repos have a subscription from arcade that is disabled. Thoughts?

**Comment by chcosta** [comment]
All of the subscription metrics we have in the PKPI epic are about "input" subscriptions.  I think we should focus on the input perspective rather than the output perspective.

As a repo owner, I want to be able to go to barviz and find how my repo is participating in shipping a particular product (channel).

> My understanding is that everything under a channel on the left hand drop down should be considered a source repo

I think we may be interpreting "source" repo differently.  Not sure I follow.

**Comment by adiaaida** [comment]
I didn't realize that about the PKPIs. 

What I meant by source repo is that they are repos that produce the dependencies that flow to other repos (source -> target). If we are focused on the input subscriptions (subscriptions that flow into the current repository), then I think the filter should filter based on the input branch, rather than the current channel or we will miss subscriptions (as I said, the arcade -> corefx subscription for corefx master is missing because it comes from the .Net Tools - Latest channel, rather than the selected .Net Core 5 Dev).

**Comment by chcosta** [comment]
I think Arcade, *will* be addressed by the other half of this PKPI which deals with "missing" subscriptions.  I'd expect that the Arcade asset would show up in the repos `eng/version.details.xml` file and we have enough info in the db to correlate an asset to a list of possible subscriptions which would provide the asset.

**Comment by chcosta** [comment]
I'm thinking, something like this

image


"Possible source" isn't the best name for "missing" though.

**Comment by adiaaida** [comment]
What are you considering "missing"?

**Comment by chcosta** [comment]
> What are you considering "missing"?

See the [doc](https://github.com/dotnet/arcade/blob/master/Documentation/ProductKPIs.md)

File: src/Maestro/maestro-angular/src/app/page/build/build.component.ts (hunk 1 above)
Comment (meganaquinn) [reply src/Maestro/maestro-angular/src/app/page/build/build.component.ts]: That is an excellent question best answered by someone who will actually be using it. :) I'll follow up on this.

File: src/Maestro/maestro-angular/src/app/page/build/build.component.ts (hunk 2 above)
Comment (meganaquinn) [reply src/Maestro/maestro-angular/src/app/page/build/build.component.ts]: Yes, it's future work if it's necessary - that information comes from either a very complicated build graph query (which would require that I pull in the build graph to this component) or it comes from VersionDetails, which I'm in the process of loading anyway. The PKPIs split product vs toolset, so I assumed it was an important part of the metric but it's good to double check.

**Comment by meganaquinn** [comment]
First shot at the UI for missing/extra subs. Does it make more sense to have it this way or under the status tab like above?

File: src/Maestro/maestro-angular/src/app/page/build/build.component.ts (hunk 1 above)
Comment (meganaquinn) [reply src/Maestro/maestro-angular/src/app/page/build/build.component.ts]: @mmitche Can you weigh in on this? Are toolsets an important part of this view and should they be toggled separately from the dependency view?

**Comment by adiaaida** [comment]
I think the dependencies without subscriptions makes sense, but the subscriptions with no dependencies should just be part of the original table, where the "Enabled" column becomes a "Status" column, with options like Enabled, Disabled, and No Dependencies

**Comment by chcosta** [comment]
Thanks for the screenshot, that's really useful.  I'm assuming that the original view just shows the "Source repository" section, and then you click the "Load Additional details" button and the other sections show up?  That seems reasonable to me. 

Is "Number of dependencies referenced" an interesting metric? Listing the dependencies might be more interesting because then the repo owner can determine whetehr the subscription is needed or not without looking at their repo.  It's not a big deal to add that feature though, I'd think just listing the repo / branch would be sufficient.

I assume that moving the branch info out of the tab and into the table's is still to be done.  I'm imagining something like...

```
https://github.com/dotnet/sdk (master)
```

**Comment by meganaquinn** [comment]
The existing subscriptions load on page load, then anything that needs the VersionDetails (so missing/extra subs as well as which subs are toolset or product) loads when "Load Additional Details" is clicked.

It's easy to list off the repos, I have all the data, but there are a lot in some cases (most are one or two, but I know I saw a 57). Would you want it as an expanding table?

File: src/Maestro/maestro-angular/src/app/page/subscriptions-table/subscriptions-table.component.html (hunk 5 above)
Comment (meganaquinn) [reply src/Maestro/maestro-angular/src/app/page/subscriptions-table/subscriptions-table.component.html]: If you don't want tabs, how do you want it handled for cases where the same subscription is referenced from multiple branches? Multiple rows?

File: src/Maestro/maestro-angular/src/app/page/build/build.component.html (hunk 3 above)
Comment (meganaquinn) [reply src/Maestro/maestro-angular/src/app/page/build/build.component.html]: Agreed - it'll either be in or out.

File: src/Maestro/maestro-angular/src/app/page/subscriptions-table/subscriptions-table.component.html (hunk 5 above)
Comment (chcosta) [reply src/Maestro/maestro-angular/src/app/page/subscriptions-table/subscriptions-table.component.html]: That was my thinking.  Are there situations where that would be weird?

**Comment by chcosta** [comment]
Having a button for loading VersionDetails seems fine to me

File: src/Maestro/maestro-angular/src/app/page/build/build.component.ts [hunk 6]
```diff
@@ -138,10 +142,10 @@ export class BuildComponent implements OnInit, OnChanges {
     } | undefined = undefined;
     const buildId$ = params$.pipe(
       filter(params => {
-        if(prevParams) {
+        if (prevParams) {
```
[the remark this answers is not in the document]
Comment (meganaquinn) [reply src/Maestro/maestro-angular/src/app/page/build/build.component.ts]: It's not my code, so I'd rather not mess with the logic. The only change here is an automatic formatting change done by VS.

File: src/Maestro/maestro-angular/src/app/page/subscriptions-table/subscriptions-table.component.html (hunk 5 above)
Comment (meganaquinn) [reply src/Maestro/maestro-angular/src/app/page/subscriptions-table/subscriptions-table.component.html]: My assumption was that the subs were edited per branch, which makes the tab grouping a more logical way to view them, but I will admit that I don't really know what would make the most sense for repo owners in this case.

Comment (meganaquinn) [reply src/Maestro/maestro-angular/src/app/page/subscriptions-table/subscriptions-table.component.html]: Something more like this?

image
image

File: src/Maestro/maestro-angular/src/app/page/build/build.component.ts (hunk 1 above)
Comment (meganaquinn) [reply src/Maestro/maestro-angular/src/app/page/build/build.component.ts]: Resolving this comment since it was split into a separate issue & pulled out of this PR.

File: src/Maestro/maestro-angular/src/app/page/build/build.component.ts (hunk 2 above)
Comment (meganaquinn) [reply src/Maestro/maestro-angular/src/app/page/build/build.component.ts]: Moving the toolsets question out into a separate issue/PR so this no longer applies.

File: src/Maestro/maestro-angular/src/app/page/subscriptions-table/subscriptions-table.component.html (hunk 5 above)
Comment (meganaquinn) [reply src/Maestro/maestro-angular/src/app/page/subscriptions-table/subscriptions-table.component.html]: Or maybe "Output Branch" instead of "Branch" to be extra clear?

Comment (chcosta) [reply src/Maestro/maestro-angular/src/app/page/subscriptions-table/subscriptions-table.component.html]: Under the first section, for "No dependencies".  I'd expect there to be branch information still in that scenario, and then the "no dependencies" info to be in the "Status" column.  Perhaps that's what you mean by "misaligned"?  If so, great, glad we're on the same page.

For the "Dependencies Without Subscriptions" section, do we have the dependency info?  ie, could we make the table look something like this?

| Dependency without a subscription | Source repository |
| --------------------------------------- | ------------------- |
| System.Foo.dll                                   | https://github.com/dotnet/corefx |
| System.Fun.dll                                   | https://github.com/dotnet/corefx |

It would probably help to run this by @mmitche and get his thoughts on what would be the most useful view here.

In general, this is getting close.  Really appreciate the screenshots.

Comment (meganaquinn) [reply src/Maestro/maestro-angular/src/app/page/subscriptions-table/subscriptions-table.component.html]: Yes on the first point - it just needed the branch column added.

I can do the Dependencies Without Subscriptions by individual dependency, but it'll be long (1-59 lines per missing subscription). I'm worried it's going to be too long, especially alongside the conflicting subscriptions work I did last night. If you think that information is going to be something that people want then we might want to have those two under optionally expanding sections that can be collapsed. (I'm saying this with no idea if that's possible or how to do it, just with the assumption that because I've seen it on other websites that it exists.)

image

Comment (meganaquinn) [reply src/Maestro/maestro-angular/src/app/page/subscriptions-table/subscriptions-table.component.html]: image

image

image

Comment (chcosta) [reply src/Maestro/maestro-angular/src/app/page/subscriptions-table/subscriptions-table.component.html]: I don't think the "dependencies without subscriptions" section is needed.  

With regards to the dependencies, how do you know the branch?  I don't think you can since multiple branches could provide the dependency and it's up to the repo owner to figure out which one is needed.

"Show / Hide Dependencies" doesn't seem like the right label because it implies (to me), that I'm going to get information about all dependencies (not just dependencies missing subscriptions), but that's not what this table is about.  Also, it occurs to me that there is an overload of the term "dependencies".  In version.details.xml, dependencies references the assets desired.  In the db, dependencies refers to the repo, not the asset and the items are called assets.  So maybe "show / hide assets without subscriptions"?

What would it look like if a different repo started providing an asset?  ie, if a different repo starts producing an asset then you might have to decide to get an asset from dotnet/repoA@master or dotnet/repoB@release/3.1.

There's enough concrete visualization here to have a meaningful conversation with Mitchell.  Sync with him, summarize the conversation outcome here, and then let's get in a room and agree upon the desired result rather than forcing you to go back and forth on these iterations.  Sound reasonable?

**Comment by meganaquinn** [comment]
Per the discussion today I stripped out everything that's going to require the new logic. Right now it's just a list of subscriptions with no data from the Version.Details.xml or other sources.

image

File: src/Maestro/Maestro.Web/Controllers/AzDevController.cs [hunk 7]
```diff
@@ -46,5 +47,20 @@ public async Task<IActionResult> GetBuildStatus(string account, string project,
                         Convert.ToBase64String(Encoding.UTF8.GetBytes(":" + token)));
                 });
         }
+
+        [HttpGet("getFile/{account}/{project}/{repo}/{*filePath}")]
```
Comment (adiaaida) [inline src/Maestro/Maestro.Web/Controllers/AzDevController.cs]: I assume that this code is getting the Versions.Details file? Or is it doing something else?

Comment (meganaquinn) [reply src/Maestro/Maestro.Web/Controllers/AzDevController.cs]: It is and it should have been pulled out with the other stuff that does that. Thanks for the catch! :)

File: src/Maestro/maestro-angular/src/app/page/build/build.component.html [hunk 8]
```diff
@@ -101,18 +101,22 @@ <h5 class="card-title float-left">
         <ng-container *stateful="let graph from graph$; loadingTemplate: loadingData">
           <ul class="nav nav-tabs">
             <li class="nav-item">
-              <a [routerLink]="['..', 'graph']" routerLinkActive="active" style="cursor: pointer;" class="nav-link">Graph</a>
+              <a [routerLink]="['..', 'graph']" routerLinkActive="active" style="cursor: pointer;"
```
Comment (chcosta) [inline src/Maestro/maestro-angular/src/app/page/build/build.component.html]: Is this just formatting changes?  If this is unrelated to your goal, can you revert this so the changes in this PR are more scoped?  Maybe clean this up separately (if it's needed, I'm not sure what the coding convention is here).

File: src/Maestro/Maestro.Web/Controllers/AzDevController.cs [hunk 9]
```diff
@@ -4,6 +4,7 @@
 using System.Text;
 using System.Threading.Tasks;
 using Maestro.AzureDevOps;
+using Microsoft.AspNetCore.Authorization;
```
Comment (chcosta) [inline src/Maestro/Maestro.Web/Controllers/AzDevController.cs]: Is this needed?

File: src/Maestro/maestro-angular/src/app/page/build/build.component.ts (hunk 1 above)
Comment (chcosta) [inline src/Maestro/maestro-angular/src/app/page/build/build.component.ts]: Is the `includeSubToolsets` variable needed for this PR?

**Comment by chcosta** [comment]
How has this been validated?  ie, even if it's as simple as picking something like "core-sdk" in the ".NET Core 5 Dev" channel and pasting the subscriptions which are shown with your changes, that would allow someone to manually validate the subscription list is correct.

File: src/Maestro/maestro-angular/src/app/page/build/build.component.html (hunk 8 above)
Comment (meganaquinn) [reply src/Maestro/maestro-angular/src/app/page/build/build.component.html]: These are all changes that will happen again if someone runs auto-format on the file. No idea if the defaults in code are correct for it though, so I rolled them back.

**Comment by meganaquinn** [comment]
I can load up the website to pull the data that way, but I don't have darc commands working, is there somewhere else I can do a manual comparison or should I find someone to run the darc command for me?

**Comment by meganaquinn** [comment]
Confirmed the data with a spot check in darc.

**Comment by chcosta** [comment]
Fantastic!  Can you include the test case info here so someone else can repro your validation?

**Comment by meganaquinn** [comment]
Sure thing - I compared the list between the Subscriptions box on the page and the results of 'darc get-subscriptions' (heavily filtered in notepad++ since the darc command orders by input repo and the page is by output repo). I spot checked three or four repos across a couple channels.
